File: src/main/scala/aes/scala/AesCryptUtil.scala
package aes.scala

import java.io.IOException
import java.nio.file.{Files, Paths}
import java.security.{InvalidAlgorithmParameterException, InvalidKeyException, NoSuchAlgorithmException, SecureRandom}
import java.util.Base64
import javax.crypto.spec.{IvParameterSpec, SecretKeySpec}
import javax.crypto.{BadPaddingException, Cipher, IllegalBlockSizeException, NoSuchPaddingException}

object AesCryptUtil {

  val IV_LENGTH = 16

  def readKey(filePath: String): Array[Byte] = {
    var key = Array.empty[Byte]

    try {
      key = Files.readAllBytes(Paths.get(filePath))
    } catch {
      case e: IOException => println(e.getMessage)
    }
    key
  }

  def encrypt(plainText: String, key: Array[Byte]): String = {
    var encryptIVAndText: String = ""

    var plainTextBytes = plainText.getBytes

    // Generate random IV
    val ivParameterSpec = createIV()
    val secretKeySpec = new SecretKeySpec(key, "AES")
    try {
      val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
      cipher.init(Cipher.ENCRYPT_MODE, secretKeySpec, ivParameterSpec)
      val encrypted = cipher.doFinal(plainTextBytes)
      // Combine IV and encrypt part
      val encryptedIVAndTextBytes = new Array[Byte](IV_LENGTH + encrypted.length)
      val iv = ivParameterSpec.getIV
      Array.copy(iv, 0, encryptedIVAndTextBytes, 0, IV_LENGTH)
      Array.copy(encrypted, 0, encryptedIVAndTextBytes, IV_LENGTH, encrypted.length)
      encryptIVAndText = Base64.getEncoder.encodeToString(encryptedIVAndTextBytes)
    } catch {
      case e@(_: NoSuchAlgorithmException | _: NoSuchPaddingException | _
        : InvalidKeyException | _: InvalidAlgorithmParameterException | _
                : IllegalBlockSizeException | _: BadPaddingException
        ) => println(e.getMessage)
    } finally {
      plainTextBytes = null //delete plainText

    }

    return encryptIVAndText
  }

  def decrypt(encryptIVAndText: String, key: Array[Byte]): String = {
    var plainText = ""

    // Extract Iv
    val encryptedIVAndTextBytes = Base64.getDecoder.decode(encryptIVAndText)
    val iv = new Array[Byte](IV_LENGTH)
    Array.copy(encryptedIVAndTextBytes, 0, iv, 0, iv.length)
    val ivParameterSpec = new IvParameterSpec(iv)

    // Extract encrypt part
    val encrytedSize = encryptedIVAndTextBytes.length - IV_LENGTH
    val encryptedBytes = new Array[Byte](encrytedSize)

    Array.copy(encryptedIVAndTextBytes, IV_LENGTH, encryptedBytes, 0, encrytedSize)

    val secretKeySpec = new SecretKeySpec(key, "AES")
    var plainTextBytes = Array.empty[Byte]
    try {
      val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
      cipher.init(Cipher.DECRYPT_MODE, secretKeySpec, ivParameterSpec)
      plainTextBytes = cipher.doFinal(encryptedBytes)
      plainText = new String(plainTextBytes)
    } catch {
      case e@(_: NoSuchAlgorithmException | _: NoSuchPaddingException | _
        : InvalidKeyException | _: InvalidAlgorithmParameterException | _
                : IllegalBlockSizeException | _: BadPaddingException
        ) => println(e.getMessage)
    } finally {
      plainTextBytes = null
    }

    return plainText
  }

  def createIV(): IvParameterSpec = {
    val iv = new Array[Byte](IV_LENGTH)
    val secureRandom = new SecureRandom()
    secureRandom.nextBytes(iv)
    new IvParameterSpec(iv)
  }

  def main(args: Array[String]): Unit = {
    val key = readKey("key.bin")
    val testText = "hello, world"
    val cipherText = encrypt(testText, key)
    val plainText = decrypt(cipherText, key)
    assert(plainText == testText)
  }

}

File: src/main/scala/aes/java/AesCryptUtil.java
package aes.java;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Base64;

public class AesCryptUtil {

    // Random IV length
    private static final int IV_LENGTH = 16;

    private static byte[] readKey(String keyPath) {
        byte[] key = null;

        try {
            key = Files.readAllBytes(Paths.get(keyPath));
        } catch (IOException e) {
            System.err.println(e.getMessage());
        }
        return key;
    }

    // Base64.encode(IV + cipherText)
    public static String encrypt(String plainText, byte[] key) {
        String encryptIVAndText = null;

        byte[] plainTextBytes = plainText.getBytes();

        // Generate random IV
        IvParameterSpec ivParameterSpec = createIv();
        SecretKeySpec secretKeySpec = new SecretKeySpec(key, "AES");
        byte[] encrypted = null;
        try {
            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            cipher.init(Cipher.ENCRYPT_MODE, secretKeySpec, ivParameterSpec);
            encrypted = cipher.doFinal(plainTextBytes);

            // Combine IV and encrypt part
            byte[] encryptedIVAndTextBytes = new byte[IV_LENGTH + encrypted.length];
            byte[] iv = ivParameterSpec.getIV();
            System.arraycopy(iv, 0, encryptedIVAndTextBytes, 0, IV_LENGTH);
            System.arraycopy(encrypted, 0, encryptedIVAndTextBytes, IV_LENGTH, encrypted.length);
            encryptIVAndText = Base64.getEncoder().encodeToString(encryptedIVAndTextBytes);

        } catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException |
                InvalidAlgorithmParameterException | IllegalBlockSizeException | BadPaddingException e) {
            System.err.println(e.getMessage());
        } finally {
            plainTextBytes = null; //delete plainText
        }

        return encryptIVAndText;
    }

    public static String decrypt(String encryptIVAndText, byte[] key) {
        String plainText = null;

        // Extract IV
        byte[] encryptedIVAndTextBytes = Base64.getDecoder().decode(encryptIVAndText);
        byte[] iv = new byte[IV_LENGTH];
        System.arraycopy(encryptedIVAndTextBytes, 0, iv, 0, iv.length);
        IvParameterSpec ivParameterSpec = new IvParameterSpec(iv);

        // Extract encrypt part
        int encrytedSize = encryptedIVAndTextBytes.length - IV_LENGTH;
        byte[] ciperTextBytes = new byte[encrytedSize];

        System.arraycopy(encryptedIVAndTextBytes, IV_LENGTH, ciperTextBytes, 0, encrytedSize);

        SecretKeySpec secretKeySpec = new SecretKeySpec(key, "AES");
        byte[] plainTextBytes = null;
        try {
            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            cipher.init(Cipher.DECRYPT_MODE, secretKeySpec, ivParameterSpec);
            plainTextBytes = cipher.doFinal(ciperTextBytes);
            plainText = new String(plainTextBytes);
        } catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException |
                InvalidAlgorithmParameterException | IllegalBlockSizeException | BadPaddingException e) {
            System.err.println(e.getMessage());
        } finally {
            plainTextBytes = null;
        }

        return plainText;
    }

    private static IvParameterSpec createIv() {
        byte[] iv = new byte[IV_LENGTH];
        SecureRandom secureRandom = new SecureRandom();
        secureRandom.nextBytes(iv);
        return new IvParameterSpec(iv);
    }

    public static void main(String[] args) {
        byte[] key = readKey("key.bin");
        String testText = "hello, world";
        String cipherText = AesCryptUtil.encrypt(testText, key);
        String plainText = AesCryptUtil.decrypt(cipherText, key);

        assert (plainText.equals(testText));
    }

}
